File: code/42.cpp
class Solution {
public:
    int trap(vector<int>& height) {
        int n = height.size();
        int best_left[n];
        int left = 0;
        int right = 0;
        int best_right[n];
        for (int i = 0; i < n; i++) {
            if (height[i] > left) {
                left = height[i];
            }
            if (height[n-i-1] > right) {
                right = height[n-i-1];
            }
            best_left[i] = left;
            best_right[n-i-1] = right;
        }
        int tot = 0;
        int water = 0;
        for (int j = 0; j < n; j++) {
            water = min(best_left[j], best_right[j]);
            tot += water - height[j];
        }
        return tot;        
    }
};

File: code/42.py
class Solution:
    def trap(self, height: List[int]) -> int:
        best_left = []
        best_right = []
        bl = 0
        br = 0
        n = len(height)
        for i in range(n):
            bl = max(bl, height[i])
            br = max(br, height[n-i-1])
            best_left.append(bl)
            best_right.append(br)
        best_right = best_right[::-1]
        rain = 0
        for i in range(n):
            rain += min(best_left[i], best_right[i]) - height[i]
        return rain
